packages/cli/src/cmds/compare/DiffLoader.ts

```diff
@@ -0,0 +1,125 @@
+
+    const roots = [...sourcePathRoots].sort().join(' ');
+
+    this.diffQueue.push(roots);
+    await this.diffQueue.drain();
+
+    return this.diffByRoots.get(roots);
+  }
+
+  private async executeDiffCommand(roots: string) {
+    if (roots === '') return;
+
+    if (this.diffByRoots.get(roots)) return;
+
+    const command = `git diff ${this.baseRevision}..${this.headRevision} -- ${roots}`;
+    let diff: Diff = { command, files: [] };
+    try {
+      const diffStr = await executeCommand(command, true);
+      if (diffStr.length > MAX_DIFF_LENGTH) {
+        warn(`Diff is too large to parse. Skipping...`);
+      } else {
+        const files = parseDiff(diffStr);
+        diff = { command, files };
+      }
+    } catch (err) {
+      // There's no point in trying this command again.
+      warn(err);
+    } finally {
+      this.diffByRoots.set(roots, diff);
+    }
+  }
+}
+
+export default class DiffLoader {
+  private sourcePathRoots = new Set<string>();
+  private diffByFile: Map<string, Diff> | undefined;
+  private diffLoaderQueue: DiffLoaderQueue;
+
+  constructor(public baseRevision: string, public headRevision: string) {
+    this.diffLoaderQueue = new DiffLoaderQueue(baseRevision, headRevision);
+  }
+
+  lookupDiff(path: string): string | undefined {
+    if (!this.diffByFile) return;
+
+    const diff = this.diffByFile.get(path);
+    if (!diff) return;
+
+    const { files: diffFiles } = diff;
+    const content = diffFiles
+      .map((diffFile) => {
+        const fromLine = ['---', diffFile.from].join(' ');
+        const toLine = ['+++', diffFile.to].join(' ');
+        const content = diffFile.chunks
+          .map((chunk) =>
+            [chunk.content, chunk.changes.map((change) => change.content).join('\n')].join('\n')
+          )
+          .join('\n');
+        return [fromLine, toLine, content, ''].join('\n');
+      })
+      .join('\n');
+    return [content].join('\n');
+  }
+
+  async update(sourcePathRoots: Set<string>) {
+    const diffRoots = new Set<string>();
+    for (const root of sourcePathRoots) {
+      if (!this.sourcePathRoots.has(root)) {
+        this.sourcePathRoots.add(root);
+        if (DiffLoader.isEligibleFile(root)) diffRoots.add(root);
+      }
+    }
+
+    const diff = await this.diffLoaderQueue.diff(diffRoots);
+    if (diff) this.parseDiffStr(diff);
+  }
+
+  static isEligibleFile(root: string): boolean {
+    return !['vendor', 'node_modules'].includes(root) && existsSync(root);
+  }
+
+  private parseDiffStr(diff: Diff) {
+    if (!this.diffByFile) this.diffByFile = new Map<string, Diff>();
+
+    diff.files.reduce((memo, diffFile) => {
+      [...new Set([diffFile.to, diffFile.from])]
+        .filter(Boolean)
+        .sort()
+        .forEach((path) => {
+          assert(path);
+          if (!memo.has(path)) memo.set(path, { command: diff.command, files: [] });
+
+          memo.get(path)!.files.push(diffFile);
+        });
+      return memo;
+    }, this.diffByFile);
+  }
```

Comment:
Using `reduce` here is quite convoluted. How about using a simple loop like this:
```suggestion
    for (const file of diff.files)
      for (const path of [file.to, file.from]) {
        if (!path) continue;
        const entry = this.diffByFile.get(path);
        if (entry) entry.files.push(file);
        else this.diffByFile.set(path, { command: diff.command, files: [file] });
      }
```
Note I removed the Set, can they really be the same? And what was the `sort` even for?

Reply:
To make the diff command stable.

Reply:
Yes `to` and `from` are the same when the file is modified, so actually they are often the same. `git diff` prints from and to like `a/app/models/user.rb`, `b/app/models.user.rb` - I am not sure why, but these are reported by parse-diff simply as `app/models/user.rb`, which makes a lot more sense to me.